ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java

```diff
@@ -30,46 +32,68 @@
+                results.put(automatonState.get(), new PerimeterOutput(preContingencyResult, automatonResult));
+                results.put(contingencyScenario.getCurativeState(), new PerimeterOutput(automatonResult, postContingencyResults.get(contingencyScenario.getCurativeState())));
+            } else {
+                results.put(contingencyScenario.getCurativeState(), new PerimeterOutput(preContingencyResult, postContingencyResults.get(contingencyScenario.getCurativeState())));
+            }
+        });
+        return results;
     }
 
     @Override
     public ComputationStatus getComputationStatus() {
         if (initialResult.getSensitivityStatus() == FAILURE
                 || postPreventiveResult.getSensitivityStatus() == FAILURE
-                || postCurativeResults.values().stream().anyMatch(perimeterResult -> perimeterResult.getSensitivityStatus() == FAILURE)) {
+                || postContingencyResults.values().stream().anyMatch(perimeterResult -> perimeterResult.getSensitivityStatus() == FAILURE)) {
             return FAILURE;
         }
         // TODO: specify the behavior in case some perimeter are FALLBACK and other ones DEFAULT
         return ComputationStatus.DEFAULT;
     }
 
     public PerimeterResult getPerimeterResult(OptimizationState optimizationState, State state) {
-
-        if (optimizationState == OptimizationState.INITIAL) {
-            if (state.getInstant() == Instant.PREVENTIVE) {
-                return null;
-            } else {
-                return postPreventiveResult;
-            }
+        if (state.getInstant().comesBefore(optimizationState.getFirstInstant())) {
+            throw new FaraoException(String.format("Trying to access results for instant %s at optimization state %s is not allowed", state.getInstant(), optimizationState));
         }
-        if (optimizationState == OptimizationState.AFTER_PRA) {
-            return postPreventiveResult;
-        }
-        if (optimizationState == OptimizationState.AFTER_CRA) {
-            if (state.getInstant() == Instant.PREVENTIVE) {
-                throw new FaraoException("Trying to access preventive results after cra is forbidden. Either get the preventive results after PRA, or the curative results after CRA.");
-            } else {
-                return postCurativeResults.get(state);
-            }
+        switch (optimizationState) {
+            case INITIAL:
+                throw new FaraoException("No PerimeterResult for INITIAL optimization state");
+            case AFTER_PRA:
+                return postPreventiveResult;
+            case AFTER_ARA:
+                if (state.getInstant().equals(Instant.CURATIVE)) {
+                    Optional<State> autoState = postContingencyResults.keySet().stream()
+                        .filter(optimizedState -> optimizedState.getInstant().equals(Instant.AUTO) && optimizedState.getContingency().equals(state.getContingency()))
+                        .findFirst();
```

Comment:
findAny() is more efficient than findFirst()

Reply:
changed